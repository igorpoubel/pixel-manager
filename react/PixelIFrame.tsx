import React, { Component } from 'react'
import { RuntimeContext, withRuntimeContext } from 'vtex.render-runtime'

import { ContextType, Pixel, PixelData } from './PixelContext'

interface Props {
  pixel: string
  runtime: RuntimeContext
}

// internal: the apps bellow need special
// access and are trusted.
const WHITELIST = [
  'vtex.request-capture-app',
]

class PixelIFrame extends Component<Props & ContextType> {
  private frame: React.RefObject<HTMLIFrameElement> = React.createRef()
  private unsubscribe?: () => void

  private pixelEventHandler = (event: string) => (data: PixelData) => {
    if (this.frame.current === null || this.frame.current.contentWindow === null) {
      return
    }

    const { runtime: { culture: { currency } } } = this.props

    const eventName = `vtex:${event}`

    const eventData = {
      currency,
      eventName,
      ...data,
    }

    this.sendEvent(this.frame.current.contentWindow, eventData)
  }

  /* tslint:disable member-ordering */
  public homeView = this.pixelEventHandler('homeView')
  public productView = this.pixelEventHandler('productView')
  public categoryView = this.pixelEventHandler('categoryView')
  public departmentView = this.pixelEventHandler('departmentView')
  public internalSiteSearchView = this.pixelEventHandler('internalSiteSearchView')
  public otherView = this.pixelEventHandler('otherView')
  public pageInfo = this.pixelEventHandler('pageInfo')
  public pageView = this.pixelEventHandler('pageView')
  public addToCart = this.pixelEventHandler('addToCart')
  public removeFromCart = this.pixelEventHandler('removeFromCart')

  public componentDidMount() {
    this.unsubscribe = this.props.subscribe(this)
  }

  public componentWillUnmount() {
    if (!this.unsubscribe) {
      return
    }

    this.unsubscribe()
  }

  public shouldComponentUpdate(prevProps: Props) {
    return prevProps.pixel !== this.props.pixel
  }

  public render() {
    const { pixel } = this.props
    const [appName] = pixel.split('@')

    return (
      <iframe
        hidden
        sandbox={WHITELIST.includes(appName) ? undefined : 'allow-scripts'}
        src={`/tracking-frame/${pixel}`}
        ref={this.frame}
      />
    )
  }
  /* tslint:enable member-ordering */

  private sendEvent = (frameWindow: Window, data: PixelData) => {
    frameWindow.postMessage(data, '*')
  }
}

export default Pixel(withRuntimeContext(PixelIFrame))
