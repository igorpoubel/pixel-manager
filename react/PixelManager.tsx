import React, { Fragment } from 'react'
import { ChildDataProps, DataProps, graphql } from 'react-apollo'

import PixelIFrame from './PixelIFrame'
import installedPixelsQuery from './queries/installedPixelsQuery.gql'

interface Data {
  installedPixels: string[]
}

const withInstalledPixels = graphql<{}, Data>(installedPixelsQuery)

const PixelManager = withInstalledPixels(({ data }) => {
  if (!data || data.loading || !data.installedPixels) {
    return null
  }

  return (
    <Fragment>
      {data.installedPixels.map(pixel => <PixelIFrame key={pixel} pixel={pixel} />)}
    </Fragment>
  )
})

export default PixelManager
