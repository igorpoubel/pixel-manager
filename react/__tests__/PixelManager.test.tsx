import { mount } from 'enzyme'
import React from 'react'
import { MockedProvider } from 'react-apollo/test-utils'
import wait from 'waait'

import PixelIFrame from '../PixelIFrame'
import PixelManager from '../PixelManager'
import installedPixelsQuery from '../queries/installedPixelsQuery.gql'

describe('<PixelManager />', () => {
  it('should render 2 iframes', async () => {
    const mocks = [
      {
        request: {
          query: installedPixelsQuery,
        },
        result: {
          data: {
            installedPixels: [
              'vtex.facebook-pixel@1.0.2',
              'vtex.google-tag-manager@1.0.0',
            ],
          },
        },
      },
    ]

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <PixelManager />
      </MockedProvider>
    )

    await wait(0)

    wrapper.update()

    expect(wrapper.find(PixelIFrame).length).toBe(2)
    expect(wrapper.find('PixelManager')).toMatchSnapshot()
  })
})
