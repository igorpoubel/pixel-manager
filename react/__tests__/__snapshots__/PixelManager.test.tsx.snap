// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PixelManager /> should render 2 iframes 1`] = `
<PixelManager
  data={
    Object {
      "error": undefined,
      "fetchMore": [Function],
      "installedPixels": Array [
        "vtex.facebook-pixel@1.0.2",
        "vtex.google-tag-manager@1.0.0",
      ],
      "loading": false,
      "networkStatus": 7,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {},
    }
  }
>
  <withPixel(PixelIFrame)
    key="vtex.facebook-pixel@1.0.2"
    pixel="vtex.facebook-pixel@1.0.2"
  >
    <PixelIFrame
      pixel="vtex.facebook-pixel@1.0.2"
      push={[Function]}
      subscribe={[Function]}
    >
      <iframe
        hidden={true}
        sandbox="allow-scripts"
        src="/tracking-frame/vtex.facebook-pixel@1.0.2"
      />
    </PixelIFrame>
  </withPixel(PixelIFrame)>
  <withPixel(PixelIFrame)
    key="vtex.google-tag-manager@1.0.0"
    pixel="vtex.google-tag-manager@1.0.0"
  >
    <PixelIFrame
      pixel="vtex.google-tag-manager@1.0.0"
      push={[Function]}
      subscribe={[Function]}
    >
      <iframe
        hidden={true}
        sandbox="allow-scripts"
        src="/tracking-frame/vtex.google-tag-manager@1.0.0"
      />
    </PixelIFrame>
  </withPixel(PixelIFrame)>
</PixelManager>
`;
